rs/registry/helpers/src/firewall.rs

```diff
@@ -116,28 +127,67 @@ impl<T: RegistryClient + ?Sized> FirewallRegistry for T {
             .flatten()
             .collect();
 
-        let mut system_subnet_endpoints: HashSet<IpAddr> = HashSet::new();
+        self.get_ip_addresses_for_node_ids(version, system_subnet_node_ids)
+    }
+
+    /// Get the IP addresses of all app subnet nodes in the registry, for endpoints used for core protocol services (p2p, xnet, api)
+    fn get_app_subnet_nodes_ip_addresses(
+        &self,
+        version: RegistryVersion,
+    ) -> RegistryClientResult<Vec<IpAddr>> {
+        let all_subnet_ids = self.get_subnet_ids(version)?.unwrap_or_default();
+        let system_subnet_ids = self.get_system_subnet_ids(version)?.unwrap_or_default();
+
+        let app_subnet_ids: Vec<SubnetId> = all_subnet_ids
+            .into_iter()
+            .filter(|id| !system_subnet_ids.contains(id))
+            .collect();
+
+        let app_subnet_node_ids: Vec<NodeId> = app_subnet_ids
+            .into_iter()
+            .map(
+                |subnet_id| match self.get_node_ids_on_subnet(subnet_id, version) {
+                    Ok(Some(node_ids)) => Ok(node_ids),
+                    Ok(None) => Ok(vec![]),
+                    Err(e) => Err(e),
+                },
+            )
+            .collect::<Result<Vec<Vec<NodeId>>, _>>()?
+            .into_iter()
+            .flatten()
+            .collect();
+
+        self.get_ip_addresses_for_node_ids(version, app_subnet_node_ids)
+    }
+
+    /// Get the IP addresses of the specified nodes
+    fn get_ip_addresses_for_node_ids(
+        &self,
+        version: RegistryVersion,
+        node_ids: Vec<NodeId>,
```

Comment:
nit: we probably don't need to take the ownership
```suggestion
        node_ids: &[NodeId],
```

Reply:
Done